zebrad/src/components/sync/status.rs

```diff
@@ -47,11 +111,52 @@ impl SyncStatus {
 
         Ok(())
     }
+
+    fn has_enough_ready_peers(&self) -> bool {
+        if self.min_ready_peers == 0 {
+            return true;
+        }
+
+        let status = (*self.peer_status.borrow()).clone();
+        status.ready_peers >= self.min_ready_peers
+    }
+
+    fn observe_tip_progress(&self) {
+        if let Some((height, _)) = self.latest_chain_tip.best_tip_height_and_hash() {
+            let mut tip = self
+                .tip_progress
+                .lock()
+                .expect("tip progress mutex poisoned");
+
+            if tip.last_height != Some(height) {
+                tip.last_height = Some(height);
+                tip.last_updated = Instant::now();
+                tip.has_progress = true;
+            }
+        }
+    }
+
+    /// Returns `true` if we've observed any activity (either a non-zero batch or a
+    /// tip height change) within the configured grace window.
+    fn recent_progress_within_grace(&self) -> bool {
+        let now = Instant::now();
+        let tip = self
+            .tip_progress
+            .lock()
+            .expect("tip progress mutex poisoned");
+        tip.has_progress && now.duration_since(tip.last_updated) <= self.progress_grace
+    }
 }
 
 impl ChainSyncStatus for SyncStatus {
     /// Check if the synchronization is likely close to the chain tip.
     fn is_close_to_tip(&self) -> bool {
+        if !self.has_enough_ready_peers() {
```

Comment:
It's okay if there aren't any ready peers, we just want to make sure that there are connected peers, ready or unready, meaning peers that recently responded to a request or sent us a request. Peer connections have an associated heartbeat task that pings them every 59s (`HEARTBEAT_INTERVAL`), and the peer set will remove peers that don't respond to it from the list of ready or unready services in the following call to its `poll_ready()` method.